### DISCUSSION TITLE
RadioButton Unchecked Event Missing

### DISCUSSION THREAD

**Discussion by usefulBeeing** [dotnet/maui#24086] [Ideas]
Unlike WPF [RadioButton](https://learn.microsoft.com/en-us/dotnet/api/system.windows.controls.radiobutton?view=windowsdesktop-8.0#events), MAUI's [RadioButton](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.controls.radiobutton?view=net-maui-8.0#events) lacks an Unchecked event. Instead, the [CheckedChanged](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.controls.radiobutton.checkedchanged?view=net-maui-8.0#microsoft-maui-controls-radiobutton-checkedchanged) event appears to handle both underlaying events. This might be fine in Code Behind, but in MVVM? I'm not sure, maybe due to lack of samples or documentation.

I would like to see CheckedChanged split into Checked and Unchecked events.

- First, this will align with the separation of concerns principle.

- Second, clear any confusion about how this 2 in 1 CheckedChanged event is supposed to work,
                especially in an MVVM scenario (Using Commands for example).

**Comment by stephenquan** [comment]
You can solve this with a TwoWay binding to IsChecked, either:

 - Subclass RadioButton with your own component and add event handlers for true/false cases as appropriate, or
 - TwoWay bind IsChecked directly to your view model, and, optionally, look at property change events for any additional handling needed in your view model

```c#
// MyRadioButton.cs
public class MyRadioButton : RadioButton
{
    public event EventHandler Checked;
    public event EventHandler Unchecked;
    public MyRadioButton()
    {
        PropertyChanged += (s,e) => {
            if (e.PropertyName == nameof(IsChecked))
            {
                if (IsChecked)
                    Checked?.Invoke(this, EventArgs.Empty);
                else
                    Unchecked?.Invoke(this, EventArgs.Empty);
            }
        };
    }
}
```

```c#
// MainViewModel.cs
public partial class MainViewModel : ObservableObject
{
     [ObservableProperty]
     bool _isChecked;

     public MainViewModel()
     {
        PropertyChanged += (s,e) =>
        {
            if (e.PropertyName == nameof(IsChecked))
            {
                if (IsChecked)
                    Debug.WriteLine("Checked");
                else
                    Debug.WriteLine("Unchecked");
            }
        };
     }
}
```

**Reply by usefulBeeing** [reply]
@stephenquan , thank you for sharing your solutions. I think they are awesome workarounds. I wonder what is your thought about the CheckedChanged event, and do you think it was not intended to be used in MVVM scenarios?

**Reply by stephenquan** [reply]
@usefulBeeing I don't mind the creation of CheckedChanged, Checked, and Unchecked events, however, from an MVVM standpoint, you should be more interested in the IsChecked property since that can achieve a more integration with your view model.

**Reply by usefulBeeing** [reply]
@stephenquan , I currently bind to RadioButtonGroup.SelectedValue, which is technically the same idea. The main reason I needed a Checked event is because I wanted to take advantage of the [AsyncRelayCommand.IsRunning](https://learn.microsoft.com/en-us/dotnet/api/communitytoolkit.mvvm.input.asyncrelaycommand?view=win-comm-toolkit-dotnet-7.0#properties) Property while using the CommunityToolkit EventToCommandBehavior. For example, to use the IsRunning property in XAML triggers or to control ActivityIndicator/s ..etc
